gmso/atomtyping/parameterize.py

```diff
@@ -0,0 +1,275 @@
+from foyer.smarts import SMARTS
+from foyer.topology_graph import TopologyGraph
+
+from gmso import Atom
+from gmso.atomtyping.isomorph import (  # check refs
+    partition_isomorphic_topology_graphs,
+)
+
+__all__ = ["apply"]
+
+
+def apply(
+    top,
+    forcefields,
+    identify_connected_components=True,
+    use_residue_info=False,
+    assert_bond_params=True,
+    assert_angle_params=True,
+    assert_dihedral_params=True,
+    assert_improper_params=False,
+):
+    """Set Topology parameter types from GMSO Forcefields.
+
+    Parameters
+    ----------
+    top: gmso.core.topology.Topology, required
+        The GMSO topology on which to apply forcefields
+    forcefields: dict, required
+        The keys are labels that match the subtopology label or site residue_name, and the
+        values are gmso Forcefield objects that gets applied to the specified molecule
+    identify_connected_components: bool, optional, default=True
+        A flag to determine whether or not to search the topology for repeated disconnected
+        structures, otherwise known as molecules and type each molecule only once.
+    use_residue_info: bool, optional, default=False
+        A flag to determine whether or not to look at site.residue_name to look parameterize
+        each molecule only once. Will only be used if identify_connected_components=False
+    assert_bond_params : bool, optional, default=True
+        If True, Foyer will exit if parameters are not found for all system
+        bonds.
+    assert_angle_params : bool, optional, default=True
+        If True, Foyer will exit if parameters are not found for all system
+        angles.
+    assert_dihedral_params : bool, optional, default=True
+        If True, Foyer will exit if parameters are not found for all system
+        proper dihedrals.
+    assert_improper_params : bool, optional, default=False
+        If True, Foyer will exit if parameters are not found for all system
+        improper dihedrals.
+    """
+    top.identify_connections()
+    top_graph = get_modified_topology_graph(top)
+    if identify_connected_components:  # True, (True,False)
+        # populate molecule information
+        isomorphic_top_graphs = partition_isomorphic_topology_graphs(top_graph)
+        apply_atomtypes_from_isomorphic(
+            top, isomorphic_top_graphs, forcefields
+        )  # what if only one forcefield?, traversing done in this function
+    elif not use_residue_info:  # False, False
+        # get entire topgraph
+        topgraph = get_modified_topology_graph(top)
+        # get rules provider from gmso_ff
+        # Assumes forcefields is just one forcefield
+        at_rules_provider = get_atomtyping_rules_provider(gmso_ff=forcefields)
+        # create a typemap of that topgraph
+        typemap = find_atomtypes(topgraph, at_rules_provider)
+        # apply typemap
+        traverse_typemap(top, typemap, forcefields)
```

Comment:
I don't think this will work for a dictionary of forcefields with more than one key right?